File: cpp/samples/10_update_service.cpp
/**
 * Sample 10 — UpdateService: firmware inventory and SimpleUpdate
 *
 * Demonstrates:
 *   - update.list_firmware_inventory()
 *   - update.get_firmware_component()
 *   - update.simple_update() and receiving a RedfishTask
 *   - Task polling via poll_task()
 *
 * NOTE: SimpleUpdate triggers an actual firmware update.  Use a simulator
 *       or non-production BMC.  Pass --dry-run to skip the update call.
 *
 * Usage:
 *   ./10_update_service [host] [port] [image-uri] [--dry-run]
 */
#include "redfish_sdk/client.hpp"
#include "redfish_sdk/errors.hpp"
#include "redfish_sdk/protocol/task.hpp"
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

// Simple arg helper
static bool has_flag(int argc, char* argv[], const std::string& flag) {
    for (int i = 1; i < argc; ++i)
        if (std::string(argv[i]) == flag) return true;
    return false;
}

int main(int argc, char* argv[]) {
    std::string host      = (argc > 1 && argv[1][0] != '-') ? argv[1] : "127.0.0.1";
    int         port      = (argc > 2 && argv[2][0] != '-') ? std::stoi(argv[2]) : 8000;
    std::string image_uri = (argc > 3 && argv[3][0] != '-')
                                ? argv[3]
                                : "http://fileserver.example.com/firmware.bin";
    bool        dry_run   = has_flag(argc, argv, "--dry-run");

    redfish::Credentials     creds{"admin", "admin"};
    redfish::ConnectionConfig config;
    config.verify_tls = false;  // disable cert verification for simulator (self-signed)
    for (int i = 1; i < argc; ++i)
        if (std::string(argv[i]) == "--no-tls") config.use_tls = false;  // plain HTTP

    try {
        auto ctx = redfish::connect(host, port, creds,
                                    redfish::AuthMode::SESSION, config);
        auto& update = ctx->update();

        // ── Firmware inventory ────────────────────────────────────────
        std::cout << "Firmware inventory:\n";
        auto fw_resp = update.list_firmware_inventory();
        std::string first_target;
        if (fw_resp.success) {
            auto members = fw_resp.body.value("Members",
                                              nlohmann::json::array());
            std::cout << "  Items: " << members.size() << "\n";
            for (const auto& item : members) {
                std::string uri = item.value("@odata.id", "");
                auto detail = ctx->get(uri);
                if (detail.success) {
                    auto& b = detail.body;
                    std::cout << "    "
                              << std::left << std::setw(20)
                              << b.value("Id","?")
                              << std::setw(30)
                              << b.value("Name","")
                              << " v" << b.value("Version","?")
                              << "\n";
                }
                if (first_target.empty()) first_target = uri;
            }
        } else {
            std::cout << "  ✗ HTTP " << fw_resp.status_code << "\n";
        }

        // ── SimpleUpdate ──────────────────────────────────────────────
        std::vector<std::string> targets;
        if (!first_target.empty()) targets.push_back(first_target);

        if (dry_run) {
            std::cout << "\n[DRY RUN] Skipping SimpleUpdate call\n";
        } else {
            std::cout << "\nCalling SimpleUpdate:\n";
            std::cout << "  ImageURI : " << image_uri << "\n";
            std::cout << "  Targets  : ";
            for (const auto& t : targets) std::cout << t << " ";
            std::cout << "\n";

            try {
                auto resp = update.simple_update(image_uri, "HTTP", targets);
                if (resp.success) {
                    std::string task_uri = resp.location();
                    if (task_uri.empty() && resp.body.is_object())
                        task_uri = resp.body.value("@odata.id", "");

                    if (!task_uri.empty()) {
                        std::cout << "  ✓ Task created: " << task_uri << "\n";
                        std::cout << "  Monitoring task (30s timeout) …\n";

                        // Poll the task
                        auto final_resp = redfish::poll_task(
                            ctx->http_client(),
                            ctx->auth_state(),
                            task_uri,
                            /*poll_interval_sec=*/2,
                            /*timeout_sec=*/30,
                            [](const redfish::RedfishTask& t) {
                                std::cout << "    state=" << t.messages
                                          << "  " << t.percent_complete
                                          << "%\n";
                            }
                        );
                        std::cout << "  ✓ Task finished  HTTP "
                                  << final_resp.status_code << "\n";
                    } else {
                        std::cout << "  ✓ Update completed synchronously"
                                  << " (HTTP " << resp.status_code << ")\n";
                    }
                } else {
                    std::cout << "  ✗ SimpleUpdate HTTP "
                              << resp.status_code << "\n";
                    std::cout << "    (Simulator may not implement SimpleUpdate)\n";
                }
            } catch (const std::exception& ex) {
                std::cout << "  ✗ SimpleUpdate error: " << ex.what() << "\n";
            }
        }

        std::cout << "\n✓ Update service sample complete\n";
        return 0;
    } catch (const redfish::RedfishError& e) {
        std::cerr << "[ERROR] " << e.what() << "\n";
    }
    return 1;
}

File: python/samples/10_update_service.py
# SPDX-License-Identifier: MIT
# Copyright (c) Microsoft Corporation. All rights reserved.

#!/usr/bin/env python3
"""Sample 10 — UpdateService: firmware/software inventory and SimpleUpdate.

Demonstrates:
  - ctx.update.list_firmware_inventory_async()
  - ctx.update.list_software_inventory_async()
  - ctx.update.simple_update_async()
  - Receiving and monitoring a RedfishTask from the update

NOTE: SimpleUpdate issues an actual firmware update command.  Use the
      simulator or a non-production BMC to avoid unintended side-effects.
      Pass --dry-run to skip the actual update call.

Usage:
    python 10_update_service.py [--host HOST] [--port PORT] [--dry-run]
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import redfish_sdk
from redfish_sdk import AuthMode, ConnectionConfig, Credentials
from redfish_sdk.errors import RedfishConnectionError, RedfishTaskFailedError


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Sample 10 — Update Service")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--user", default="admin")
    p.add_argument("--password", default="password")
    p.add_argument("--no-tls-verify", action="store_true")
    p.add_argument("--no-tls", action="store_true", help="Use plain HTTP instead of HTTPS (for simulators without SSL)")
    p.add_argument(
        "--image-uri",
        default="http://fileserver.example.com/firmware.bin",
        help="URI of the firmware image for SimpleUpdate",
    )
    p.add_argument(
        "--dry-run",
        action="store_true",
        help="Skip the actual SimpleUpdate call",
    )
    return p.parse_args()


async def main() -> None:
    args = parse_args()

    creds = Credentials(username=args.user, password=args.password)
    config = ConnectionConfig(verify_tls=not args.no_tls_verify, use_tls=not args.no_tls)

    try:
        ctx = await redfish_sdk.connect_async(
            host=args.host,
            port=args.port,
            credentials=creds,
            auth_mode=AuthMode.SESSION,
            config=config,
        )
    except RedfishConnectionError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    async with ctx:
        update = ctx.update

        # ── Firmware inventory ───────────────────────────────────────────
        print("Firmware inventory:")
        fw_resp = await update.list_firmware_inventory_async()
        if fw_resp.success:
            members = fw_resp.body.get("Members", [])
            print(f"  Items: {len(members)}")
            for item in members:
                uri = item.get("@odata.id", "")
                detail = await ctx.get_async(uri)
                if detail.success:
                    b = detail.body
                    print(
                        f"    {b.get('Id','?'):<20} "
                        f"{b.get('Name',''):<30} "
                        f"v{b.get('Version','?')}"
                    )
        else:
            print(f"  ✗ HTTP {fw_resp.status_code}")

        # ── Software inventory ───────────────────────────────────────────
        print("\nSoftware inventory:")
        sw_resp = await update.list_software_inventory_async()
        if sw_resp.success:
            members = sw_resp.body.get("Members", [])
            print(f"  Items: {len(members)}")
            for item in members:
                uri = item.get("@odata.id", "")
                detail = await ctx.get_async(uri)
                if detail.success:
                    b = detail.body
                    print(
                        f"    {b.get('Id','?'):<20} "
                        f"{b.get('Name',''):<30} "
                        f"v{b.get('Version','?')}"
                    )
        else:
            print(f"  ✗ HTTP {sw_resp.status_code}")

        # ── UpdateTargets — collect first FW target ──────────────────────
        target_uri: str | None = None
        if fw_resp.success:
            members = fw_resp.body.get("Members", [])
            if members:
                target_uri = members[0].get("@odata.id")

        # ── SimpleUpdate ─────────────────────────────────────────────────
        if args.dry_run:
            print("\n[DRY RUN] Skipping SimpleUpdate call")
        else:
            targets = [target_uri] if target_uri else []
            print(f"\nCalling SimpleUpdate:")
            print(f"  ImageURI  : {args.image_uri}")
            print(f"  Targets   : {targets}")
            print(f"  ApplyTime : Immediate")

            try:
                task = await update.simple_update_async(
                    image_uri=args.image_uri,
                    targets=targets,
                    transfer_protocol="HTTP",
                    apply_time="Immediate",
                )

                if task is None:
                    print("  ✓ Update completed synchronously (200/204)")
                else:
                    print(f"  ✓ Task created: {task.task_uri}")
                    print(f"  Task state    : {task.state}")

                    # Monitor with progress
                    print("  Monitoring task (30s timeout) …")
                    try:
                        async for snapshot in task.monitor_async():
                            pct = snapshot.percent_complete or 0
                            state = snapshot.state
                            print(f"    {state:<20} {pct}%")
                    except RedfishTaskFailedError as exc:
                        print(f"  ✗ Task failed: {exc.task.state}")
                    else:
                        print(f"  ✓ Task finished: {task.state}")

            except Exception as exc:  # noqa: BLE001
                print(f"  ✗ SimpleUpdate error: {exc}")
                print("    (Simulator may not implement SimpleUpdate)")

        print("\n✓ Update service sample complete")


if __name__ == "__main__":
    asyncio.run(main())
